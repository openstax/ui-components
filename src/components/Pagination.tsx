import React from 'react';
import styled from 'styled-components';
import { palette } from "../theme/palette";

export const LinkForPage = styled(({ page, current, href, onClick, className }: {
  page: number;
  current?: boolean;
  href?: string;
  className?: string;
  onClick?: (event: React.MouseEvent<HTMLAnchorElement>) => void;
}) => {
  const currentValue = current ? "page" : undefined;

  return (
    <a
      className={className}
      aria-current={currentValue}
      href={href || '#'}
      onClick={onClick}
    >
      {page}
    </a>
  );
})`
`;

export const Pagination = styled((props: {
  className?: string;
  Page: (props: {page: number; current: boolean}) => React.ReactElement;
  currentPage: number;
  totalPages: number;
  totalItems?: number;
  pageSize?: number;
  showFromEnd?: number;
  showFromCurrent?: number;
}) => {
  const {
    showFromEnd,
    showFromCurrent,
    pageSize,
    totalItems,
    className,
    currentPage,
    totalPages,
    Page,
  } = {
    showFromEnd: 3,
    showFromCurrent: 2,
    ...props
  };

  // the paginator would be empty, so short-circuit
  if (totalPages === 0 || totalPages === 1) {
    return null;
  }

  // prevent nav from changing size as you switch pages
  const minEntries = showFromEnd * 2 + showFromCurrent * 2 +
    1 + // the current page
    2 // for the ellipsis
  ;

  const middleRange: [number, number] = [
    Math.max(1, Math.min(currentPage - showFromCurrent, totalPages + 1)),
    Math.min(totalPages, currentPage + showFromCurrent) + 1
  ];
  const startRange: [number, number] = [
    1,
    Math.min(middleRange[0], showFromEnd + 1)
  ];
  const endRange: [number, number] = [
    Math.max(1, middleRange[1], totalPages - showFromEnd + 1),
    totalPages + 1
  ];

  const numberOfEntries = Math.max(0, startRange[1] - startRange[0]) +
    Math.max(0, middleRange[1] - middleRange[0]) +
    Math.max(0, endRange[1] - endRange[0]) +
    (startRange[1] === middleRange[0] ? 0 : 1) +
    (middleRange[1] === endRange[0] ? 0 : 1)
  ;

  if (numberOfEntries < minEntries) {
    let remaining = minEntries - numberOfEntries;
    const delta = Math.floor(remaining / 2);

    const firstGap = middleRange[0] - startRange[1];
    const secondGap = endRange[0] - middleRange[1];

    const firstMod = Math.min(firstGap, secondGap === 0
      // there is no second gap, try use entire diff in the first
      ? remaining
      : secondGap < (remaining - delta)
        // there is a gap but its smaller than the delta, so use it all
        // in the first and add one for losing the ellipsis
        ? remaining - secondGap + 1
        : delta
    );
    remaining -= firstMod;
    const secondMod = Math.min(secondGap, remaining);

    middleRange[0] = Math.max(1, middleRange[0] - firstMod);
    middleRange[1] = Math.min(totalPages + 1, middleRange[1] + secondMod);
    startRange[1] = Math.min(middleRange[0], showFromEnd + 1);
    endRange[0] = Math.max(middleRange[1], totalPages - showFromEnd + 1);
  }

  return (
    <div className={className}>
      <nav aria-label="pagination links">
        <ul>
          {range(...startRange).map((p) =>
            <li key={p} className={currentPage === p ? 'active' : undefined}>
              <Page page={p} current={currentPage === p} />
            </li>
          )}
          {startRange[1] !== middleRange[0] ?
            <li className="disabled">
              <span>...</span>
            </li>
          : null}
          {range(...middleRange).map((p) =>
            <li key={p} className={currentPage === p ? 'active' : undefined}>
              <Page page={p} current={currentPage === p} />
            </li>
          )}
          {middleRange[1] !== endRange[0] ?
            <li className="disabled">
              <span>...</span>
            </li>
          : null}
          {range(...endRange).map((p) =>
            <li key={p} className={currentPage === p ? 'active' : undefined}>
              <Page page={p} current={currentPage === p} />
            </li>
          )}
        </ul>
      </nav>
      {pageSize && totalItems ? <div className="pagination-info">
        {(currentPage - 1) * pageSize + 1}-{Math.min(currentPage * pageSize, totalItems)} of {totalItems}
      </div> : null}
    </div>
  );
})`
  text-align: center;

  > nav > ul {
    list-style: none;
    padding: 0;
    border: thin solid ${palette.neutralLight};
    border-radius: 0.5rem;
    display: inline-block;
    margin: 0 auto;

    > li {
      margin: 0;
      min-width: 4rem;
      text-align: center;
      display: inline-block;

      &:not(:last-child) {
        border-right: thin solid ${palette.neutralLight};
      }

      &.active,
      &:focus-within:not(.disabled),
      &:hover:not(.disabled) {
        background-color: ${palette.neutralLighter};
      }

      > ${LinkForPage},span {
        padding: 1rem;
        display: block;
        text-decoration: none;
        font-size: 1.6rem;
        line-height: 1.3rem;
        margin: 0;
        color: inherit;
      }
    }
  }

  .pagination-info {
    margin-top: 0.5rem;
    font-size: 1.6rem;
  }
`;

function range(lower: number, upper: number) {
  if (upper < lower) return [];
  return Array.from({length: upper-lower}).map((_, i) => i + lower);
}
