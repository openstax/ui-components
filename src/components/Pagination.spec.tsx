import { render } from '@testing-library/react';
import { Pagination, LinkForPage } from "./Pagination";

describe('Pagination', () => {
  let root: HTMLElement;

  beforeEach(() => {
    root = document.createElement('main');
    root.id = 'root';
    document.body.append(root);
  });

  it('matches snapshot', () => {
    render(<Pagination
      currentPage={1} totalPages={10}
      Page={({ page, current }) =>
        <LinkForPage page={page} current={current} href="#" />
      }
    />, {container: root});
    expect(document.body).toMatchSnapshot();
  });

  it('matches snapshot with dividers', () => {
    render(<Pagination
      currentPage={5} totalPages={10} showFromEnd={1} showFromCurrent={1}
      Page={({ page, current }) =>
        <LinkForPage page={page} current={current} href="#" />
      }
    />, {container: root});
    expect(document.body).toMatchSnapshot();
  });

  it('grows to min size', () => {
    render(<Pagination
      currentPage={1} totalPages={10} showFromEnd={1} showFromCurrent={1}
      Page={({ page, current }) =>
        <LinkForPage page={page} current={current} href="#" />
      }
    />, {container: root});
    expect(document.body).toMatchSnapshot();
  });

  it('grows to min size from back', () => {
    render(<Pagination
      currentPage={10} totalPages={10} showFromEnd={1} showFromCurrent={1}
      Page={({ page, current }) =>
        <LinkForPage page={page} current={current} href="#" />
      }
    />, {container: root});
    expect(document.body).toMatchSnapshot();
  });

  it('noops', () => {
    render(<Pagination
      currentPage={1} totalPages={1}
      Page={({ page, current }) =>
        <LinkForPage page={page} current={current} href="#" />
      }
    />, {container: root});
    expect(document.body).toMatchSnapshot();
  });
});
