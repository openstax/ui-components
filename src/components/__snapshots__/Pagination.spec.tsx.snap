// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pagination grows to min size 1`] = `
<body>
  <main
    id="root"
  >
    <div
      class="sc-gsnTZi Nkuun"
    >
      <nav
        aria-label="pagination links"
      >
        <ul>
          <li
            class="active"
          >
            <a
              aria-current="page"
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              1
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              2
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              3
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              4
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              5
            </a>
          </li>
          <li
            class="disabled"
          >
            <span>
              ...
            </span>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              10
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </main>
</body>
`;

exports[`Pagination grows to min size from back 1`] = `
<body>
  <main
    id="root"
  >
    <div
      class="sc-gsnTZi Nkuun"
    >
      <nav
        aria-label="pagination links"
      >
        <ul>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              1
            </a>
          </li>
          <li
            class="disabled"
          >
            <span>
              ...
            </span>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              6
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              7
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              8
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              9
            </a>
          </li>
          <li
            class="active"
          >
            <a
              aria-current="page"
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              10
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </main>
</body>
`;

exports[`Pagination matches snapshot 1`] = `
<body>
  <main
    id="root"
  >
    <div
      class="sc-gsnTZi Nkuun"
    >
      <nav
        aria-label="pagination links"
      >
        <ul>
          <li
            class="active"
          >
            <a
              aria-current="page"
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              1
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              2
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              3
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              4
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              5
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              6
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              7
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              8
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              9
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              10
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </main>
</body>
`;

exports[`Pagination matches snapshot with dividers 1`] = `
<body>
  <main
    id="root"
  >
    <div
      class="sc-gsnTZi Nkuun"
    >
      <nav
        aria-label="pagination links"
      >
        <ul>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              1
            </a>
          </li>
          <li
            class="disabled"
          >
            <span>
              ...
            </span>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              4
            </a>
          </li>
          <li
            class="active"
          >
            <a
              aria-current="page"
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              5
            </a>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              6
            </a>
          </li>
          <li
            class="disabled"
          >
            <span>
              ...
            </span>
          </li>
          <li>
            <a
              class="sc-bczRLJ kkAHbA"
              href="#"
            >
              10
            </a>
          </li>
        </ul>
      </nav>
    </div>
  </main>
</body>
`;

exports[`Pagination noops 1`] = `
<body>
  <main
    id="root"
  />
</body>
`;
